### PR TITLE
metrics: jaro distance metric

### PR DESCRIPTION
**Pull request by petros-ioannidis**
* A new file was created that will be used for string distance metrics.
* Substring_occurences was created that returns a generator with the
occurences of a string.
* Jaro matching was implemented to calculate the matching characters
  based on the Jaro metric algorithm
* Tests were developed for the aforementioned features

### CODE REVIEW

File: beard/metrics/string_distance.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Beard.
+# Copyright (C) 2014, 2015 CERN.
```
Comment (glouppe) [inline beard/metrics/string_distance.py]: 2015 only

File: beard/metrics/string_distance.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
+# -*- coding: utf-8 -*-
```
Comment (glouppe) [inline beard/metrics/string_distance.py]: Can you rename this file to simply `string.py`? The `_distance` suffix is redundant with the fact that the file is already within a folder named "metrics".

File: beard/metrics/string_distance.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+#
+# This file is part of Beard.
+# Copyright (C) 2014, 2015 CERN.
+#
+# Beard is a free software; you can redistribute it and/or modify it
+# under the terms of the Revised BSD License; see LICENSE file for
+# more details.
+
+"""String distances metrics.
```
Comment (glouppe) [inline beard/metrics/string_distance.py]: Remove the "distances" term, as this is not strictly correct.

File: beard/metrics/string_distance.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        substring we are searching for
+
+    Returns
+    -------
+    :returns: generator
+        a generator that holds the value of the index
+    """
+    if letter == '':
+        raise ValueError('Empty string is not allowed as a parameter for letter')
+    try:
```
Comment (glouppe) [inline beard/metrics/string_distance.py]: I would remove this `try/except` and directly start from `index = 0` in the loop below. There is no need to check that `S` supports `find` (garbage in, garbage out).

File: beard/metrics/string_distance.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+# under the terms of the Revised BSD License; see LICENSE file for
+# more details.
+
+"""String distances metrics.
+
+.. codeauthor:: Petros Ioannidis <petros.ioannidis91@gmail.com>
+
+"""
+def substring_occurences(s, letter):
+    """Return the number of matching letters based on the jaro algorithm.
```
Comment (glouppe) [inline beard/metrics/string_distance.py]: I would rephrase this as 
```
def find_all(s, pattern):
    """Find all occurences of the given pattern."""
```

File: beard/metrics/string_distance.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    if letter == '':
+        raise ValueError('Empty string is not allowed as a parameter for letter')
+    try:
+        index = s.find(letter)
+    except AttributeError:
+        raise TypeError('S must be a string')
+
+    while index != -1:
+        yield index
+        index = s.find(letter, index + 1)
```
Comment (glouppe) [inline beard/metrics/string_distance.py]: By the way, couldnt we simply do instead : 

```
for match in re.finditer(pattern, s):
    yield match.start()
```

File: beard/metrics/string_distance.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+def jaro_matching(s1, s2):
+    """Return the number of matching letters based on the jaro algorithm.
```
Comment (glouppe) [inline beard/metrics/string_distance.py]: I would rephrase this as "Return the number of matching letters in s1 and s2".

File: beard/metrics/tests/test_string_distance.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+"""Test string distances metrics.
+
+.. codeauthor:: Petros Ioannidis <petros.ioannidis91@gmail.com>
+
+"""
+import pytest
+from ..string_distance import substring_occurences, jaro_matching
+
+
+@pytest.mark.parametrize('s, letter, occur',
```
Comment (glouppe) [inline beard/metrics/tests/test_string_distance.py]: Can you import `parametrize` above to remove some boilerplate code?

**Comment by glouppe** [comment]
Since this is still quite short, I will wait for you to implement the full Jaro distance function before merging this.

File: beard/metrics/tests/test_string_distance.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
+# -*- coding: utf-8 -*-
```
Comment (glouppe) [inline beard/metrics/tests/test_string_distance.py]: Same here, rename the file to `test_string.py`

File: beard/metrics/tests/test_string_distance.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+# # This file is part of Beard.
+# Copyright (C) 2014, 2015 CERN.
```
Comment (glouppe) [inline beard/metrics/tests/test_string_distance.py]: 2015 only

File: beard/metrics/tests/test_string_distance.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+#
+# Beard is a free software; you can redistribute it and/or modify it
+# under the terms of the Revised BSD License; see LICENSE file for
+# more details.
+
+"""Test string distances metrics.
```
Comment (glouppe) [inline beard/metrics/tests/test_string_distance.py]: without "distances"

File: beard/metrics/string_distance.py (hunk 2 above)
Comment (petros-ioannidis) [inline beard/metrics/string_distance.py]: There is a `string` package in python so I suggest we try to find another name.

File: beard/metrics/string_distance.py (hunk 6 above)
Comment (petros-ioannidis) [inline beard/metrics/string_distance.py]: You are right. I runned timeit and the code you proposed runs faster. I will change it.

File: beard/metrics/string_distance.py (hunk 2 above)
Comment (petros-ioannidis) [inline beard/metrics/string_distance.py]: I used `string_methods.py` for a temporary name until we find something more meaningful.

File: beard/metrics/string_distance.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        The number of matching letters
+    """
+
+    H = min(len(s1), len(s2))//2
+
+    letters_cache = dict()
+    matches = 0
+
+    for letter in s1:
+        if not letters_cache.has_key(letter):
```
Comment (MSusik) [inline beard/metrics/string_distance.py]: `in` operator does the same, is faster and is easier to read.

File: beard/metrics/string_distance.py [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    :param s1: string
+        first string
+
+    :param s2: string
+        second string
+
+    Returns
+    -------
+    :returns: int
+        The number of matching letters
```
Comment (MSusik) [inline beard/metrics/string_distance.py]: You have some inconsistency in your comments:

` a generator that holds the value of the index`
` The number of matching letters`
uppercase vs lowercase, "a" vs "the"

The trend is to use uppercase and omit articles. Maybe we should standarize it.

Comment (petros-ioannidis) [inline beard/metrics/string_distance.py]: Omitting articles is not suggested by PEP8. I fixed the uppercase.

File: beard/metrics/string_methods.py [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+    :param letter: string
+        Substring we are searching for
+
+    Returns
+    -------
+    :returns: generator
+        A generator that holds the indexes of the patterns
+    """
+    for match in re.finditer(pattern, s):
+            yield match.start()
```
Comment (glouppe) [inline beard/metrics/string_methods.py]: indentation is not consistent here

File: beard/metrics/string_methods.py [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+
+def jaro_matching(s1, s2):
+    """Return the number of matching letters based on the jaro algorithm.
```
Comment (glouppe) [inline beard/metrics/string_methods.py]: Strictly speaking, there is no "jaro algorithm", but only a "jaro similarity metric" . So I would rephrase this as "Return the number of matching letters in the two strings"

**Comment by glouppe** [comment]
For the filename, `text.py` should be fine.

File: beard/metrics/tests/test_string_methods.py [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+# under the terms of the Revised BSD License; see LICENSE file for
+# more details.
+
+"""Test string metrics.
+
+.. codeauthor:: Petros Ioannidis <petros.ioannidis91@gmail.com>
+
+"""
+import pytest
+from pytest import mark
```
Comment (glouppe) [inline beard/metrics/tests/test_string_methods.py]: Why not `from pytest.mark import parametrize`?

Comment (petros-ioannidis) [inline beard/metrics/tests/test_string_methods.py]: `mark` is not a module.

**Comment by glouppe** [comment]
@petros-ioannidis During development, can you avoid amending/squashing commits? This makes much harder to follow the discussion in github. We will squash everything before merge anyway.

File: beard/metrics/tests/test_string_methods.py (hunk 16 above)
Comment (glouppe) [inline beard/metrics/tests/test_string_methods.py]: Ok, I just realized it. (yuck)

File: beard/metrics/text.py [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+    Returns
+    -------
+    :returns: generator
+        A generator that holds the indexes of the patterns
+    """
+    for match in re.finditer(pattern, s):
+        yield match.start()
+
+
+def jaro_matching(s1, s2):
```
Comment (glouppe) [inline beard/metrics/text.py]: Can you rename this function? Since it now returns both the number of matching letters and number of transpositions. Please update the docstring as well.

File: beard/metrics/text.py [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+                    break
+    s1_matching_letter_sequence.sort()
+    s2_matching_letter_sequence.sort()
+    transpositions = len(filter(lambda x: x[0][1] != x[1][1],
+                         zip(s1_matching_letter_sequence,
+                             s2_matching_letter_sequence)))
+    return matches, transpositions
+
+
+def jaro_similarity_metric(s1, s2):
```
Comment (glouppe) [inline beard/metrics/text.py]: Simply name it `jaro`

File: beard/metrics/text.py [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+        Similarity of s1 and s2
+    """
+    s_prime = jaro_matching(s1, s2)[0]
+    s = len(s1)
+    t_prime = jaro_matching(s2, s1)[0]
+    t = len(s2)
+
+    return (1/3)*((s_prime/s) +
+                  (t_prime/t) +
+                  ((s_prime - (jaro_matching(s1, s2)[1]/2))/(s_prime)))
```
Comment (glouppe) [inline beard/metrics/text.py]: What definition of the metric are you using?

File: beard/metrics/text.py (hunk 17 above)
Comment (petros-ioannidis) [inline beard/metrics/text.py]: Yes I forgot. Should I make 2 different functions or a tuple is fine? I did it like this for 'optimization' purposes but maybe its less clear like that. What do you think?

Comment (glouppe) [inline beard/metrics/text.py]: Computing the two values at the same time is fine with me.

File: beard/metrics/clustering.py [hunk 20]
```diff
@@ -38,7 +38,7 @@ def paired_precision_recall_fscore(labels_true, labels_pred):
 
     Returns
     -------
-    :return float precision: calculated precission
+    :return float precision: calculated precision
```
Comment (etzemis) [inline beard/metrics/clustering.py]: This is an outdated version of the file.

Comment (petros-ioannidis) [inline beard/metrics/clustering.py]: I rebased, it should be fine now.

File: beard/metrics/text.py (hunk 17 above)
Comment (glouppe) [inline beard/metrics/text.py]: In the end, we dont reuse nor call this function more than once. Shouldn't we put directly the computation of the number of matching letters / transpositions directly into `jaro`?

File: beard/metrics/text.py [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+    :param s2: string
+        Second string
+
+    Returns
+    -------
+    :returns: float
+        Similarity of s1 and s2
+    """
+    s_data = jaro_matching(s1, s2)
+    s_prime = s_data[0]
```
Comment (glouppe) [inline beard/metrics/text.py]: Could you instead unpack both values directly? It would be more readable. 

E.g. `n_matches, n_transpositions = jaro_matching(s1, s2)`

File: beard/metrics/text.py [hunk 22]
```diff
@@ -0,0 +1,126 @@
...
+        Second string
+
+    Returns
+    -------
+    :returns: int
+        The number of matching letters
+    """
+    H = min(len(s1), len(s2))//2
+
+    letters_cache = dict()
```
Comment (glouppe) [inline beard/metrics/text.py]: Use `{}`, for consistency with how you initialize lists below.

File: beard/metrics/text.py [hunk 23]
```diff
@@ -0,0 +1,126 @@
...
+                if pos_s1 - H <= pos_s2 <= pos_s1 + H:
+                    if pos_s1 not in s1_matched_positions:
+                        s1_matched_positions.append(pos_s1)
+                    s2_matching_letter_sequence.append((pos_s2, letter))
+                    break
+
+    s1_matching_letter_sequence.sort()
+    s2_matching_letter_sequence.sort()
+    transpositions = len(tuple(filter(lambda x: x[0][1] != x[1][1],
+                         zip(s1_matching_letter_sequence,
```
Comment (glouppe) [inline beard/metrics/text.py]: Can you align `zip` with `filter`, to make it more readable?

File: beard/metrics/text.py [hunk 24]
```diff
@@ -0,0 +1,126 @@
...
+                             s2_matching_letter_sequence))))
+    return matches, transpositions
+
+
+def jaro(s1, s2):
+    """Return the similarity of the strings s1 and s2.
```
Comment (glouppe) [inline beard/metrics/text.py]: Return the Jaro similarity between strings `s1` and `s2`.

File: beard/metrics/text.py [hunk 25]
```diff
@@ -0,0 +1,126 @@
...
+
+    Calculate the similarity based on the Jaro similarity metric.
```
Comment (glouppe) [inline beard/metrics/text.py]: and then remove this sentence, which becomes redundant with the one above

File: beard/metrics/text.py [hunk 26]
```diff
@@ -0,0 +1,126 @@
...
+        for pos_s2 in s2_occur:
+            for pos_s1 in s1_occur:
+                if pos_s1 - H <= pos_s2 <= pos_s1 + H:
+                    if pos_s1 not in s1_matched_positions:
+                        s1_matched_positions.append(pos_s1)
+                    s2_matching_letter_sequence.append((pos_s2, letter))
+                    break
+
+    s1_matching_letter_sequence.sort()
+    s2_matching_letter_sequence.sort()
```
Comment (glouppe) [inline beard/metrics/text.py]: Hmm I feel there is a more straightforward way to compute all these values. I am not not sure the caching you do above is worth it in the end because you need to resort the sequences. I'll check this more closely tomorrow.

File: beard/metrics/text.py [hunk 27]
```diff
@@ -0,0 +1,126 @@
...
+    s2_matching_letter_sequence = []
+    s1_matched_positions = []
+    s2_matched_positions = []
+
+    for letter in s1:
+        if letter not in letters_cache:
+            letters_cache[letter] = (tuple(find_all(s1, letter)),
+                                     tuple(find_all(s2, letter)))
+
+    for letter, (s1_occur, s2_occur) in letters_cache.items():
```
Comment (MSusik) [inline beard/metrics/text.py]: `items()` has different behaviour in Python2 and Python3. If you want to be consistent, use `six` library.

http://pythonhosted.org/six/#six.iteritems

File: beard/metrics/text.py [hunk 28]
```diff
@@ -0,0 +1,126 @@
...
+        for pos_s1 in s1_occur:
+            for pos_s2 in s2_occur:
+                if pos_s1 - H <= pos_s2 <= pos_s1 + H:
+                    if pos_s2 not in s2_matched_positions:
+                        matches += 1
+                        s2_matched_positions.append(pos_s2)
+                    s1_matching_letter_sequence.append((pos_s1, letter))
+                    break
+
+    for letter, (s1_occur, s2_occur) in letters_cache.items():
```
Comment (MSusik) [inline beard/metrics/text.py]: The same as above

File: beard/metrics/tests/test_text.py [hunk 29]
```diff
@@ -0,0 +1,68 @@
...
+    assert tuple(find_all(s, letter)) == occur
+
+
+@mark.parametrize('s, letter',
+                  (('MARTHA', 'Z'),
+                   ('', 'A')))
+def test_find_all_none_string(s, letter):
+    """Test find_all behaviour for empty cases."""
+    with pytest.raises(StopIteration):
+        assert next(find_all(s, letter))
```
Comment (MSusik) [inline beard/metrics/tests/test_text.py]: Might be nice to use here `six` library:

http://docs.pythonsprints.com/python3_porting/py-porting.html#iterators
http://pythonhosted.org/six/#six.next

File: beard/metrics/text.py (hunk 27 above)
Comment (petros-ioannidis) [inline beard/metrics/text.py]: When used as an `iterable` the behaviour is the same in Python2 and Python3. So it is fine like that.

Comment (MSusik) [inline beard/metrics/text.py]: IMO, no! As a C++ programmer you should know that the type should be as strict as possible.

Anyway, let's drop the opinionable argument and let's think about the performance.
In python 2 `items()` method creates a new list of tuples. That is both time and memory consuming.
`six.iteritems()` just uses python 2 `iteritems()`.

Comment (petros-ioannidis) [inline beard/metrics/text.py]: We are not coding in `C++` though, so I suggest you focus on the pythonic way. ;)
Items is fine for this use case. If you want you can benchmark it and if it is really a blocker for performance I can change it. ;)

File: beard/metrics/text.py (hunk 17 above)
Comment (glouppe) [inline beard/metrics/text.py]: Can you make this private?

Comment (petros-ioannidis) [inline beard/metrics/text.py]: Done.

**Comment by glouppe** [comment]
Thanks for the last changes. Can you proceed and implement jaro-winkler in this same branch? (it should be quite straightforward now)

File: beard/metrics/tests/test_text.py [hunk 30]
```diff
@@ -0,0 +1,76 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Beard.
+# Copyright (C) 2015 CERN.
+#
+# Beard is a free software; you can redistribute it and/or modify it
+# under the terms of the Revised BSD License; see LICENSE file for
+# more details.
+
+"""Test string metrics.
```
Comment (glouppe) [inline beard/metrics/tests/test_text.py]: string -> text (for consistency with the file name)

File: beard/metrics/text.py [hunk 31]
```diff
@@ -0,0 +1,153 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Beard.
+# Copyright (C) 2015 CERN.
+#
+# Beard is a free software; you can redistribute it and/or modify it
+# under the terms of the Revised BSD License; see LICENSE file for
+# more details.
+
+"""String metrics.
```
Comment (glouppe) [inline beard/metrics/text.py]: string -> text

File: beard/metrics/text.py [hunk 32]
```diff
@@ -0,0 +1,153 @@
...
+    """
+    for match in re.finditer(pattern, s):
+        yield match.start()
+
+
+def _jaro_matching(s1, s2):
+    """Return the number of matching letters.
+
+    Calculate the number of matching letters
+    based on the method for computing the Jaro distance metric.
```
Comment (glouppe) [inline beard/metrics/text.py]: Can you remove this sentence? I think it is redundant with the first sentence.

File: beard/metrics/text.py [hunk 33]
```diff
@@ -0,0 +1,153 @@
...
+    -------
+    :returns: generator
+        A generator that holds the indexes of the patterns
+    """
+    for match in re.finditer(pattern, s):
+        yield match.start()
+
+
+def _jaro_matching(s1, s2):
+    """Return the number of matching letters.
```
Comment (glouppe) [inline beard/metrics/text.py]: The function is also computing the number of transpositions. Hence it should be mentioned. 

Accordingly, the function should be renamed as well.

File: beard/metrics/text.py [hunk 34]
```diff
@@ -0,0 +1,153 @@
...
+    :param s1: string
+        First string
+
+    :param s2: string
+        Second string
+
+    Returns
+    -------
+    :returns: int
+        The number of matching letters
```
Comment (glouppe) [inline beard/metrics/text.py]: Same here, please correct the docsting

File: beard/metrics/text.py (hunk 27 above)
Comment (glouppe) [inline beard/metrics/text.py]: Can you rename `s1_occur` to `s1_positions`? I think it would be clearer. The same for `s2_occur`.

File: beard/metrics/text.py [hunk 35]
```diff
@@ -0,0 +1,153 @@
...
+    s2_matched_positions = []
+
+    for letter in s1:
+        if letter not in letters_cache:
+            letters_cache[letter] = (tuple(find_all(s1, letter)),
+                                     tuple(find_all(s2, letter)))
+
+    for letter, (s1_occur, s2_occur) in letters_cache.items():
+        for pos_s1 in s1_occur:
+            for pos_s2 in s2_occur:
```
Comment (glouppe) [inline beard/metrics/text.py]: can you rename `pos_s1` to `i` and `pos_s2` to `j`?

File: beard/metrics/text.py [hunk 36]
```diff
@@ -0,0 +1,153 @@
...
+    -------
+    :returns: int
+        The number of matching letters
+    """
+    H = min(len(s1), len(s2))//2
+
+    letters_cache = {}
+    matches = 0
+    transpositions = 0
+    s1_matching_letter_sequence = []
```
Comment (glouppe) [inline beard/metrics/text.py]: `s1_matching_letters` would be a bit shorter

File: beard/metrics/text.py [hunk 37]
```diff
@@ -0,0 +1,153 @@
...
+        Second string
+
+    Returns
+    -------
+    :returns: float
+        Similarity of s1 and s2
+    """
+    n_matches, n_transpositions = _jaro_matching(s1, s2)
+    s = len(s1)
+    t = len(s2)
```
Comment (glouppe) [inline beard/metrics/text.py]: Can you remove these two variable and directly call `len(s1)` (`len(s2)`) in the expression below?

File: beard/metrics/text.py [hunk 38]
```diff
@@ -0,0 +1,153 @@
...
+
+    return (1/3)*((n_matches/s) +
+                  (n_matches/t) +
+                  ((n_matches - (n_transpositions/2))/(n_matches)))
+
+
+def jaro_winkler(s1, s2):
```
Comment (glouppe) [inline beard/metrics/text.py]: Can you instead make the scaling factor `p` a parameter of this function, with `p=0.1` by default?

File: beard/metrics/text.py [hunk 39]
```diff
@@ -0,0 +1,153 @@
...
+
+    :param s2: string
+        Second string
+
+    Returns
+    -------
+    :returns: float
+        Similarity of s1 and s2
+    """
+    _jaro = jaro(s1, s2)
```
Comment (glouppe) [inline beard/metrics/text.py]: `_jaro` -> `jaro_distance`

File: beard/metrics/text.py [hunk 40]
```diff
@@ -0,0 +1,153 @@
...
+
+    P = 0
+    for s1_letter, s2_letter in zip(s1, s2):
+        if s1_letter == s2_letter:
+            P += 1
+        else:
+            break
+
+    P_prime = min(P, 4)
```
Comment (glouppe) [inline beard/metrics/text.py]: this variable is not useful, directly use `min()` in the expression below

File: beard/metrics/text.py [hunk 41]
```diff
@@ -0,0 +1,153 @@
...
+        Second string
+
+    Returns
+    -------
+    :returns: float
+        Similarity of s1 and s2
+    """
+    _jaro = jaro(s1, s2)
+
+    P = 0
```
Comment (glouppe) [inline beard/metrics/text.py]: I would rename this as `common_prefix` to make it easier to understand.

File: beard/metrics/text.py [hunk 42]
```diff
@@ -0,0 +1,153 @@
...
+    for s1_letter, s2_letter in zip(s1, s2):
```
Comment (glouppe) [inline beard/metrics/text.py]: Note that we could stop the loop early, as soon as `P==4`.

File: beard/metrics/text.py [hunk 43]
```diff
@@ -0,0 +1,153 @@
...
+        First string
+
+    :param s2: string
+        Second string
+
+    Returns
+    -------
+    :returns: float
+        Similarity of s1 and s2
+    """
```
Comment (glouppe) [inline beard/metrics/text.py]: Can you add a reference towards the original paper?

File: beard/metrics/text.py [hunk 44]
```diff
@@ -0,0 +1,153 @@
...
```
Comment (glouppe) [inline beard/metrics/text.py]: Can you add a reference towards the original paper?

File: beard/metrics/text.py (hunk 38 above)
Comment (petros-ioannidis) [inline beard/metrics/text.py]: Wouldn't this alter the algorithm? It is specified in the description of the algorithm.

Comment (glouppe) [inline beard/metrics/text.py]: I have seen several papers where they adjust the scaling factor. I think it is best to leave open the possibility to do that as well, as long as the default value matches the default algorithm.

File: beard/metrics/tests/test_text.py [hunk 45]
```diff
@@ -0,0 +1,76 @@
...
+    """Test jaro_matching behaviour."""
+    assert _jaro_matching(s1, s2) == match
+
+
+@mark.parametrize('s1, s2, match',
+                  (('MARTHA', 'MARHTA', 0.944),
+                   ('DWAYNE', 'DUANE', 0.822)))
+def test_jaro(s1, s2, match):
+    """Test jaro_similarity_metric behaviour."""
+    assert abs(jaro(s1, s2) - match) < 0.001
```
Comment (glouppe) [inline beard/metrics/tests/test_text.py]: Can you use assert_almost_equal from numpy.testing instead?

File: beard/metrics/tests/test_text.py [hunk 46]
```diff
@@ -0,0 +1,76 @@
...
+
+
+@mark.parametrize('s1, s2, match',
+                  (('MARTHA', 'MARHTA', 0.961),
+                   ('DWAYNE', 'DUANE', 0.84)))
+def test_jaro_winkler(s1, s2, match):
+    """Test jaro_similarity_metric behaviour."""
+    assert abs(jaro_winkler(s1, s2) - match) < 0.001
```
Comment (glouppe) [inline beard/metrics/tests/test_text.py]: Same here

File: beard/metrics/text.py [hunk 47]
```diff
@@ -0,0 +1,170 @@
...
+    jaro_distance = jaro(s1, s2)
+
+    common_prefix = 0
+    for s1_letter, s2_letter in zip(s1, s2):
+        if s1_letter == s2_letter and common_prefix < 4:
+            common_prefix += 1
+        else:
+            break
+
+    return jaro_distance + (p * min(common_prefix, 4) * (1 - jaro_distance))
```
Comment (glouppe) [inline beard/metrics/text.py]: Now that the loop breaks early, you can directly use `common_prefix` instead of `min(...)`

File: beard/metrics/text.py [hunk 48]
```diff
@@ -0,0 +1,170 @@
...
+    -------
+    :returns: float
+        Similarity of s1 and s2
+
+    Reference
+    ---------
+    Jaro, M. A. 1989. Advances in record-linkage methodology as
+    applied to matching the 1985 census of Tampa, Florida.
+    Journal of the American Statistical Association
+    84:414–420.
```
Comment (glouppe) [inline beard/metrics/text.py]: This one is enough

File: beard/metrics/tests/test_text.py [hunk 49]
```diff
@@ -0,0 +1,76 @@
...
+                   ('MARHTA', 'MARTHA', (6, 2))))
+def test_jaro_matching(s1, s2, match):
+    """Test jaro_matching behaviour."""
+    assert _jaro_matching(s1, s2) == match
+
+
+@mark.parametrize('s1, s2, match',
+                  (('MARTHA', 'MARHTA', 0.944),
+                   ('DWAYNE', 'DUANE', 0.822)))
+def test_jaro(s1, s2, match):
```
Comment (glouppe) [inline beard/metrics/tests/test_text.py]: Can you test the border cases as well? ie. when strings are identical / dont share any letters / one is empty.

**Comment by glouppe** [comment]
Great, thanks for your patience. +1 for me. Can you squash?

@MSusik is it ok for you?

**Comment by MSusik** [comment]
> @MSusik is it ok for you?

IMO using `items()` is wrong. Ask @etzemis  for a review.

**Comment by petros-ioannidis** [comment]
The commits are squashed and the PR is ready to be merged.

**Comment by etzemis** [comment]
:+1: for me for merging.

**Comment by glouppe** [comment]
Thanks, merging this
